### PR TITLE
[FLINK-25445] No need to create local recovery dirs when disabled loc…

### PR DESCRIPTION
**Pull request by zuston**
…al-recovery in TaskExecutorLocalStateStoresManager

## What is the purpose of the change

`TaskExecutor` will create file, check `localRecoveryEnabled` and load local state store for each task submission in method `localStateStoreForSubtask`. For batch jobs, the `localRecoveryEnabled` is always false, and there's no need to create these local files for task in `TaskExecutor`

## Brief change log

When disable task local recovery, it will not create subtask state dirs.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zuston** [comment]
Could you help review? @dmvk Thanks

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java [hunk 1]
```diff
@@ -31,11 +29,10 @@
     private final boolean localRecoveryEnabled;
 
     /** Encapsulates the root directories and the subtask-specific path. */
-    @Nonnull private final LocalRecoveryDirectoryProvider localStateDirectories;
+    private final LocalRecoveryDirectoryProvider localStateDirectories;
 
     public LocalRecoveryConfig(
-            boolean localRecoveryEnabled,
-            @Nonnull LocalRecoveryDirectoryProvider directoryProvider) {
+            boolean localRecoveryEnabled, LocalRecoveryDirectoryProvider directoryProvider) {
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java:35]:
```suggestion
            boolean localRecoveryEnabled, @Nullable LocalRecoveryDirectoryProvider directoryProvider) {
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java [hunk 2]
```diff
@@ -44,7 +41,6 @@ public boolean isLocalRecoveryEnabled() {
         return localRecoveryEnabled;
     }
 
-    @Nonnull
     public LocalRecoveryDirectoryProvider getLocalStateDirectoryProvider() {
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java:44]: Would it make sense to change the signature to `Optional<LocalRecoveryDirectoryProvider> getLocalStateDirectoryProvider()` to make sure the change is properly reflected in all places?

Other option would be adding some explicit exception when getting a null provider.

**Review by dmvk — CHANGES_REQUESTED** [review]
Thanks for the PR @zuston! This already looks pretty good ;) I went through the usages of the `LocalRecoveryDirectoryProvider` and the change seems safe.

My only concern is around proper handling of `@Nullable` providers in `LocalRecoveryConfig`, PTAL

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java (hunk 1 above)
Comment (zuston) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java:35]: Got it

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java (hunk 2 above)
Comment (zuston) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java:44]: The signature of optional may be a better option. We could use orElseThrow to throw exceptions although it should not happen when enabled task local recovery.

**Comment by zuston** [comment]
Thanks for your review @dmvk 

By the way, why we make task local-recovery enabled defaultly? Only due to the uncompatility with unaligned checkpoint?

**Comment by dmvk** [comment]
The behavior of the local recovery highly depends on the state backend. AFAIK heap state backend still doesn't support it. For RocksDB you can find some context in https://issues.apache.org/jira/browse/FLINK-15507

**Comment by zuston** [comment]
Thank very much @dmvk . I’ll take a look

**Comment by zuston** [comment]
Updated @dmvk Could you help review again? Thanks

**Comment by zuston** [comment]
@dmvk Gentle ping. If you have time, could you help review it? At weekends, i will have more time to deal with it.
Thanks

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java [hunk 3]
```diff
@@ -44,9 +48,10 @@ public boolean isLocalRecoveryEnabled() {
         return localRecoveryEnabled;
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java:48]:
```suggestion
        return localStateDirectories != null;
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java [hunk 4]
```diff
@@ -58,4 +63,24 @@ public String toString() {
...
+
+        public NullOfDirProviderException(String message) {
+            super(message);
+        }
+
+        public NullOfDirProviderException() {
+            this(
+                    "When task local recovery is enabled, the NullPointerException of LocalRecoveryDirectoryProvider should not happen.");
+        }
+    }
```
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java:85]:
```suggestion

    public static Supplier<IllegalStateException> localRecoveryNotEnabled() {
        return () ->
                new IllegalStateException(
                        "Getting a LocalRecoveryDirectoryProvider is only supported with the local recovery enabled. This is a bug and should be reported.");
    }
```

**Review by dmvk — CHANGES_REQUESTED** [review]
Hi @zuston, I've added few more minor comments. I think this is mostly good to go.

My primary concern would be that we should get rid of the `SavepointLocalRecoveryProvider` from the state processor API, as it's no longer needed after this patch.

Thanks for working on this! 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java (hunk 4 above)
Comment (zuston) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalRecoveryConfig.java]: 👍

**Comment by zuston** [comment]
Updated @dmvk. Could you help check again? 
Thanks~

**Comment by zuston** [comment]
Gentle ping @dmvk . Thanks :)

**Comment by zuston** [comment]
Rebase latest master and squash all commits. @dmvk If you have time, could you help review again? Thanks :)

**Comment by zuston** [comment]
@flinkbot run azure

**Comment by zuston** [comment]
@dmvk Gentle ping. Could u help review it ? Thanks

**Review by dmvk — APPROVED** [review]
LGTM 👍 Thanks for the contribution

**Pull request merged**